Add some basic methods for MinimalSemigroupGeneratingSet

wilfwilson opened the pull request:
It is going to be very difficult to implement completely general methods for `MinimalSemigroupGeneratingSet` that have reasonable performance. However, I think we should still write methods that apply in special cases.

Here I've added several of those special cases. The `IndecomposableElements` attribute added in PR #347 was useful for this.

The biggest class of semigroups to which the new method applies is probably finite J-trivial semigroups.

james-d-mitchell commented on gap/attributes/attr.gi:
```diff
@@ -960,3 +960,87 @@ function(S)
+    return gens;
+  fi;
+
+  ErrorNoReturn("not yet implemented,");
```
This error message does not have the correct format.

james-d-mitchell commented on doc/attr.xml:
```diff
@@ -337,8 +337,9 @@ gap> Length(SmallGeneratingSet(S));
   <Description>
 
-    <B>Warning:</B> currently, no methods are installed to compute these
-    attributes.<P/>
```
I'm confused, this documentation existed before, but there are no methods installed? Would it not make more sense to not document (or indeed declare) attributes where there are no methods installed? Maybe I misunderstood something here, but this seems like an oddity to me.

james-d-mitchell replied:
And the example below seems to contradict the assertion that there are no methods installed for `MinimalInverseMonoidGeneratingSet`. The distinction between there being a method that "computes" a minimal generating set, and one that simply knows (like a monogenic inverse semigroup) a minimal generating set is probably too subtle. I suggest improving this warning to make it clearer what is going on.

james-d-mitchell commented:
@wilfwilson any comments on my comments?

wilfwilson commented:
Yes, I'll address it and hopefully make the changes at the next Wednesday meeting. A minimal monoid generating set is simply a minimal semigroup generating set, except you throw away the identity if present (except possibly for partial perms...). I'll have to think and talk to you about minimal inverse semigroup generating sets.

I'm happy to not document the attributes that genuinely have no methods installed. But it is useful to declare them so that we can set them when we create things that we know have minimal (whatever) generating sets: eg if we create a monogenic inverse monoid, for instance.

wilfwilson commented:
I think I've addressed your changes. I've added methods for `MinimalMonoidGeneratingSet`. We decided not to do a fallback brute-force method, and I'm not immediately should what is the right thing to do for inverse semigroups.

wilfwilson commented:
The Travis tests are failing because of the coverage job, even though it shouldn't be. I think it's a problem with the profiling package. I'll investigate.

wilfwilson commented:
I've changed Travis to use the master version of `profiling`. That seems to fix the tests. I'm happy for this to be reviewed again @james-d-mitchell and/or merged.

james-d-mitchell commented:
@wilfwilson happy to merge this if you rebase

wilfwilson commented:
Done, thanks @james-d-mitchell.
